dask/base.py

```diff
@@ -1147,105 +1147,92 @@ def normalize_random_state(state):
-    except TypeError:  # not hashable
-        return _normalize_callable(func)
-
-
-def _normalize_callable(func: Callable) -> tuple | str | bytes:
-    if isinstance(func, Compose):
-        first = getattr(func, "first", None)
-        funcs = reversed((first,) + func.funcs) if first else func.funcs
-        return tuple(normalize_callable(f) for f in funcs)
-    elif isinstance(func, (partial, curry)):
-        args = tuple(normalize_token(i) for i in func.args)
-        if func.keywords:
-            kws = tuple(
-                (k, normalize_token(v)) for k, v in sorted(func.keywords.items())
+
+@normalize_token.register(object)
+def normalize_object(o):
+    method = getattr(o, "__dask_tokenize__", None)
+    if method is not None:
+        return method()
+
+    if type(o) is object:
+        if config.get("tokenize.ensure-deterministic"):
+            raise RuntimeError(
+                "object() cannot be deterministically hashed. See "
+                "https://docs.dask.org/en/latest/custom-collections.html#implementing-deterministic-hashing "
+                "for more information."
             )
-        else:
-            kws = None
-        return normalize_callable(func.func), args, kws
-    else:
-        try:
-            result = pickle.dumps(func, protocol=4)
-            if b"__main__" not in result:  # abort on dynamic functions
-                return result
-        except Exception:
-            pass
-        if not config.get("tokenize.ensure-deterministic"):
-            try:
-                import cloudpickle
+        # Idempotent, but not deterministic. Make sure that the id is not reused.
+        _seen_objects.add(o)
+        return "object", id(o)
 
-                return cloudpickle.dumps(func, protocol=4)
-            except Exception:
-                return str(func)
-        else:
+    if dataclasses.is_dataclass(o) and not isinstance(o, type):
+        return _normalize_dataclass(o)
+
+    try:
+        return _normalize_pickle(o)
+    except Exception:
+        if config.get("tokenize.ensure-deterministic"):
             raise RuntimeError(
-                f"Function {func!r} may not be deterministically hashed by "
-                "cloudpickle. See: https://github.com/cloudpipe/cloudpickle/issues/385 "
+                f"Object {o!r} cannot be deterministically hashed. See "
+                "https://docs.dask.org/en/latest/custom-collections.html#implementing-deterministic-hashing "
                 "for more information."
             )
+        return uuid.uuid4().hex
+
+
+def _normalize_pickle(o: object) -> tuple:
+    buffers: list[pickle.PickleBuffer] = []
+    pik: bytes | None
+    try:
+        pik = pickle.dumps(o, protocol=5, buffer_callback=buffers.append)
+        if b"__main__" in pik:
+            pik = None
+    except Exception:
+        pik = None
+
+    if pik is None:
+        buffers.clear()
+        pik = cloudpickle.dumps(o, protocol=5, buffer_callback=buffers.append)
```

Comment:
`main` disallows using cloudpickle when `ensure-deterministic` is enabled. This behavior was introduced in https://github.com/dask/dask/issues/9135. I don't consider the example reported in https://github.com/dask/dask/issues/9135 sufficient motivation for this behavior but cloudpickle itself states that it is not deterministic

https://github.com/cloudpipe/cloudpickle/issues/453
https://github.com/cloudpipe/cloudpickle/issues/385

and I'm a bit on the fence about this since I do not know what to expect now.

Reply:
For instance, should this function be idempotent?

```python
import dask
from dask.base import tokenize
dask.config.set({"tokenize.ensure-deterministic": True})

def foo():
    import pickle
    class A:
        pass
    return tokenize(A())

assert foo() == foo()  # Should this be True or False?
```

For this PR this is `False` since every invocation of `foo` returns a different token. On `main` this is raising a `RuntimeError`

Reply:
The intention of the `RuntimeError` is to alert a user that a call includes a custom object that does not implement the `__dask_tokenize__` protocol, see https://github.com/dask/dask/issues/6555 for the first discussion

Reply:
For the record, this dynamic class determinism is not working with cloudpickle due to a uuid that is being used internally, see https://github.com/cloudpipe/cloudpickle/blob/d003266b18336e1e603536bdbe6518bc2dcc00d3/cloudpickle/cloudpickle.py#L112
i.e. cloudpickle is intentionally distinguishing the two calls. Do we want or need to do the same? From a dask user perspective, I think both calls are basically identical.

Maybe this doesn't matter, I don't know

Reply:
my gut tells me it is safer to raise if we are not sure. At this time it is hard to estimate the impact on dask-expr later and removing an exception later on is easier than introducing one

Reply:
It's worth noting that your example fails for local _instances_, but passes for local _classes_ and _functions_:

```python
# PASS
def test_tokenize_local_classes_from_different_contexts():
    def f():
        class C:
            pass

        return C

    assert check_tokenize(f()) == check_tokenize(f())


# FAIL
def test_tokenize_local_instances_from_different_contexts():
    def f():
        class C:
            pass

        return C()

    assert check_tokenize(f()) == check_tokenize(f())


# PASS
def test_tokenize_local_functions_from_different_contexts():
    def f():
        def g():
            return 123

        return g

    assert check_tokenize(f()) == check_tokenize(f())

```

IMHO it's not terribly common to define whole classes in a local context and I can't think of a real-life scenario where this will make things break. I'm not saying impossible, just that for me it feels like an edge case.

I've added the three tests above and xfailed the instance one, linking the upstream cloudpickle issue.

Reply:
Important to note is that the difference in token is generated by calling f() twice. If you call f() once, then pass the returned object through a cloudpickle round-trip, the token remains the same.

Reply:
Also worth noting that `assert f() == f()` fails in all three above cases. So to me in the two cases that pass we're already doing more than what the Python interpreter does natively.